src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs

```diff
@@ -404,9 +404,44 @@ internal override bool ApplyNullableTransforms(byte defaultTransformFlag, Immuta
             return true;
         }
 
-        internal override TypeSymbol SetNullabilityForReferenceTypes(Func<TypeWithAnnotations, TypeWithAnnotations> transform)
+        internal override TypeSymbol SetObliviousNullabilityForReferenceTypes()
         {
-            return WithElementType(transform(ElementTypeWithAnnotations));
+            if (ElementType.TypeKind == TypeKind.Array)
+            {
+                // The language supports deeply nested arrays, up to 10,000+ instances. This means we can't implemented a head
+                // recursive solution here. Need to take an iterative approach to building up the annotations here.
+                var builder = ArrayBuilder<ArrayTypeSymbol>.GetInstance();
+
+                // Dig through the element types to get the set of ArrayTypeSymbol instances that need to be marked
+                // as oblivious
+                var array = this;
+                TypeWithAnnotations elementType;
+                do
+                {
+                    builder.Add(array);
+                    elementType = array.ElementTypeWithAnnotations;
+                    array = elementType.TypeKind == TypeKind.Array
+                        ? (ArrayTypeSymbol)elementType.Type
+                        : null;
```

Comment:
I don't believe this is faster than `array = elementType.Type as ArrayTypeSymbol;`
In fact I bet it is slower.